okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/UrlSelectorImpl.java

```diff
@@ -103,17 +125,36 @@ private static String switchWsToHttp(String url) {
         // if possible, determine the index of the passed in url (so we can be sure to return a url which is different)
         Optional<Integer> existingUrlIndex = indexFor(existingUrl);
 
-        // ...otherwise we fall back to the stateful current url index
-        int index = currentUrl.updateAndGet(
-                (currentUrlIndex) -> (existingUrlIndex.orElse(currentUrlIndex) + 1) % baseUrls.size());
-        return redirectTo(existingUrl, baseUrls.get(index));
+        int numAttempts = 0;
+        int potentialNextIndex = existingUrlIndex.orElse(currentUrl.get());
+
+        // Find the next URL that is not marked as failed
+        while (numAttempts < baseUrls.size()) {
+            potentialNextIndex = (potentialNextIndex + 1) % baseUrls.size();
+            Boolean maybeInCache = failedUrls.getIfPresent(baseUrls.get(potentialNextIndex));
```

Comment:
Could you store a single variant enum `FAILED` instead of a boolean? It's kinda confusing that we're never reading the value of that boolean and just checking it for null!